ENH/DOC: signal.hilbert2: Refactor to avoid extra scaling array and improve docstr

JesseLivezey opened the pull request:
#### What does this implement/fix?
Similar to #23173 for `hilbert2`. This PR
- avoids allocating an extra scaling array (I get modest speedups on a laptop),
- improves the docstring ~and adds a reference to the particular definition of 2d analytic signal used here
- add an argument to allow performing batched `2d` transform on `nd` arrays for `n >= 2`

TODOs:
- [x] add a test for using non-default axes
- [x] add a direct test that `ifft2` of analytic signal has expected orthant power.
- [x] what's the right version to add for the added arg? 1.16.2 was added.
- [ ] check that docs look okay

FYI, @DietBru , similar to the `hilbert()` PR.

JesseLivezey committed:
ENH/DOC: signal.hilbert2: Refactor to avoid extra scaling array and improve docstr

j-bowhay commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -3497,21 +3499,85 @@ def test_bad_args(self, xp):
+        x = xp.arange(xp.prod(shape)).reshape(shape)
+        xh_old = _hilbert2(x)
+        xh = hilbert2(x)
+        assert_allclose(xh_old, xh)
```
```suggestion
        xp_assert_close(xh_old, xh)
```

JesseLivezey commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -3497,21 +3499,85 @@ def test_bad_args(self, xp):
+            x = sp_fft.ifft2(Xf * h, axes=(0, 1))
+            return x
+
+        x = xp.arange(xp.prod(shape)).reshape(shape)
```
```suggestion
        x = xp.arange(shape[0] * shape[1]).reshape(shape)
```

DietBru commented:
I found a little time to look at it today:
The referenced [Wikipedia article] does not give a definition but only states that there are different approaches for defining two-dimensional analytic signals.. It would be helpful, if the docstr would state how the 2d Hilbert transform / analytic signal is defined for this function. An authoritative reference would even be better.

The following example illustrates what is happening:
```python
import numpy as np
from scipy.fft import fft2, fftshift, ifftshift
from scipy.signal import hilbert2

# Input signal is unit impulse with DC offset:
x = np.ones((5, 5)) / 25
x[0, 0] += 1
X = fftshift(fft2(x))  # Zero frequency bin is at center

y = hilbert2(x)
Y = fftshift(fft2(y))

print("FFT2 of Input:")
print(X)
print("FFT2 of Output:")
print(np.round(Y, 3))
```
It produces:
```text
FFT2 of Input:
[[1.-0.j 1.-0.j 1.-0.j 1.+0.j 1.+0.j]
 [1.-0.j 1.-0.j 1.-0.j 1.+0.j 1.+0.j]
 [1.-0.j 1.-0.j 2.-0.j 1.+0.j 1.+0.j]
 [1.-0.j 1.-0.j 1.+0.j 1.+0.j 1.+0.j]
 [1.-0.j 1.-0.j 1.+0.j 1.+0.j 1.+0.j]]
FFT2 of Output:
[[ 0.+0.j  0.+0.j -0.+0.j  0.+0.j  0.+0.j]
 [ 0.+0.j  0.+0.j -0.+0.j  0.+0.j  0.+0.j]
 [ 0.+0.j  0.+0.j  2.+0.j  2.+0.j  2.+0.j]
 [ 0.+0.j  0.+0.j  2.+0.j  4.+0.j  4.+0.j]
 [ 0.+0.j  0.+0.j  2.+0.j  4.+0.j  4.+0.j]]
```
It illustrates that the nonnegative frequency bins are multiplied by either two or four. 
I am not completely sure that this result concurs to the [Wikipedia article] and I was not able to find a reference (though I did not search too long).

It would be great if in this PR or in a follow-up, the docstr could be improved.

 [Wikipedia article]: https://en.wikipedia.org/wiki/Analytic_signal#Extensions_of_the_analytic_signal_to_signals_of_multiple_variables

JesseLivezey commented:
I can add better docs for the current version and a method kwarg with an option to use the single orthant definition

DietBru commented:
I am not sure if I am understanding  your definitions correctly, @JesseLivezey: Currently the values are multiplied by four, if $\omega_x > 0 \wedge \omega_y > 0$ and multiplied by two if $\omega_{x,y}=0 \wedge \omega_{y,x}>0$.

If $H(x) := 1 + \text{sign}(x)$, then `hilbert2` would follow the convention
```
AS(f(x, y)) = InvFourier( 2 * H(w_x) * H(w_y) * Fourier(f(x, y)) )
```


> I can add better docs for the current version and a method kwarg with an option to use the single orthant definition

:+1: Improved docs would be very welcome.

JesseLivezey commented:
@DietBru correcting my previous comment. The current scipy implementation _is_ based on the "Single orthant" paper to the best of my understanding (added a reference). Roughly, it zeros out any amplitude where either of the frequencies is negative and multiplies the amplitudes of the positive-positive frequency amplitudes by 4. I've tried to clarify these details in the docstring.

As you point out, this definition has some drawbacks which are pointed out in the new third reference.

DietBru reviewed (changes requested):
The refactorization of the code is a helpful simplification. Also allowing higher dimensional arrays by adding an ´axis` parameter` is a win and follows the convention of similar functions (should be added to the release notes). 

I added some suggestions to improve the docstr and the unit tests.

JesseLivezey commented on scipy/signal/_signaltools.py:
```diff
@@ -2587,37 +2587,64 @@ def hilbert(x, N=None, axis=-1):
-    """
-    Compute the '2-D' analytic signal of `x`
+def hilbert2(x, N=None, axes=(-2, -1)):
+    """Compute the '2-D' analytic signal of `x`.
```
That's great. I've included this.

JesseLivezey commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -3497,21 +3498,108 @@ def test_bad_args(self, xp):
+        if sh1 % 2 == 0:
+            n_quads = n_quads & (freq1 != xp.min(freq1))
+        zero_quad = x_as_f[n_quads]
+        xp_assert_close(zero_quad, xp.zeros_like(zero_quad), atol=1e-12)
```
Done!

DietBru commented:
I rebased this PR onto the current branch hoping CI would pass. There's seems to be a [problem] in CuPy & friends when doing in this [test]:
```python
       ...
        x_f = xp.ones(shape, dtype=xp.complex128)  # FFT of input signal
        x_f[0 , 0] += 7
        x = xp.real(sp_fft.ifft2(x_f))  # != 0

        x_as = hilbert2(x)  # <--  == 0. (!?)
```
I do not understand why `x` became zero here (while it passes in the other tests). Do you have any insight, @j-bowhay,perhaps?

Rendered docs can be found [here].

[problem]: https://github.com/scipy/scipy/actions/runs/17317409940/job/49162890897?pr=23247#step:12:18059
[test]: https://github.com/JesseLivezey/scipy/blob/2323b0653ce037b0f8edd76ff7b3eaf78b6f5395/scipy/signal/tests/test_signaltools.py#L3412
[here]: https://output.circle-artifacts.com/output/job/62efe5e1-d68c-4026-b6c2-e0da80fc020d/artifacts/0/html/reference/generated/scipy.signal.hilbert2.html#scipy.signal.hilbert2

DietBru committed:
Attempt to make `test_signaltools.TestHilbert2` pass CuPy/Dask backend unit tests.

Also minor changes like whitespaces were reverted to keep this branch related only to `signal.hilbert2`.

Add cupy PR number to skip reason

DietBru commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -3372,31 +3372,109 @@ def test_hilbert_types(self, dtype, xp):
+        f0_pos, f0_neg = slice(1, (shape[0] + 1) // 2), slice((shape[0] + 1) // 2, None)
+        f1_pos, f1_neg = slice(1, (shape[1] + 1) // 2), slice((shape[1] + 1) // 2, None)
+        # Verify all values:
+        atol = np.finfo(x_as_f.dtype).resolution
```
Thanks for the hint. This solves only part of the problem. There seems also a CI problem–consult issue #23539.

JesseLivezey commented:
@DietBru , thanks for the changes. I'll be able to take a look in the next day or two.

JesseLivezey commented:
@DietBru , this might explain some of the discrepancies for CuPy
https://github.com/cupy/cupy/blob/f78e78e09cdb5c2d326b069e54f6e8445d7b270d/cupyx/scipy/signal/_signaltools.py#L1751
I think there is a bug in their implementation (they have `h1` twice).

DietBru reviewed (approved):
Looks good now. Thanks @JesseLivezey!

ev-br commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -3388,31 +3388,119 @@ def test_hilbert_types(self, dtype, xp):
+
+    @pytest.mark.parametrize('shape', [(4, 5), (5, 4), (4, 4), (5, 5)])
+    @skip_xp_backends("cupy", reason="Bug in cupy implementation, see cupy#9396")
+    def test_quadrant_values(self, shape, xp):
```
Here it is @JesseLivezey, as a comment to your CuPy PR:  https://github.com/cupy/cupy/pull/9396#issuecomment-3311419533

Note the process (bit convoluted, yes): 
- SciPy tests are skipped (can actually xfail them instead, but taht's minor) until the CuPy releases a version with your fix
- CuPy tests are skipped until SciPy releases 1.17 and the CuPy CI runs with scipy 1.17

Therefore, this PR can be merged at any time; CuPy PR can be merged any time; once we have the right releases, both libs can undo their skips/xfails. The last steps are for 2026, most likely :-).

JesseLivezey commented:
@DietBru, looks like things run again after another rebase. The one failure looks unrelated (I hope).

DietBru committed:
Merge branch 'main' into hilbert2-upgrades

JesseLivezey commented:
Thanks for your help @DietBru!

I figured this would be a simple repeat of the `hilbert()` PR 🤣
